### ISSUE TITLE
Implement LRP for regression

### ISSUE THREAD

**Issue by xushanthu-2014** [sebastian-lapuschkin/lrp_toolbox#23]
Hi and thanks for your work about LRP and it's a really nice work towards improving interpretability of neural network!
Currently I am looking for a suitable way to explain my model in a specific task. My task is trying to build a mapping from X to Y where Y is continuous values in stead of discrete class labels. So I am wondering if I can directly use your LRP here?

And my task is a multilabel regression which means I need to predict 3 different & independent outputs simultaneously as https://github.com/sebastian-lapuschkin/lrp_toolbox/issues/12 indicated. So how can I do that? I saw you apply a mask on the feature dimension to mask the largest value among all different features as 1 and others as 0. But I would rather keep the ground truth as its original continuous values....so how can I do that?

By the way, can I implement LRP just like SHAP? for example building model independently then shap.DeepExplainer(model, background). I saw your code that I need to rewrite my model structure and add LRP in each layer, can I just avoid rewriting my model?
Thanks and hope for your early reply!

**Comment by sebastian-lapuschkin** [comment]
Hi @xushanthu-2014 ,

1) LRP should be directly applicable to regression tasks out of the box (assuming you are not using some super-fancy NN structure with non-standard building blocks).

2) I assume you are refering to https://github.com/sebastian-lapuschkin/lrp_toolbox/blob/master/python/lrp_demo.py around line 70. Here we build the mask as a means to select single output neurons to backpropagate with LRP. We weight the network outputs with 0 and 1 respectively. Usually, one backpropagates the actual output of a neuron of interest, ie the f(x) of that neuron. There are however valid reasons to backprop normalized quantities instead, ie a 1 when the explanation should be invariant to the model's own (in)confidence in the prediction, which is required in some situations.
In the code linked above, we backprop the f(x) of the dominant network output. you can however arrange your mask however you like. Options would be: backprop each of the three output neurons separately, to understand how the model uses its input for each of the three outputs respectively. You could also backprop all outputs simultaneously to gain information about the general use of input features by the model. This will however yield harder to read heatmaps, since you mix importance information for all your outputs. It dependes on what your outputs have been trained for, and whether they are aggreeably combinable. I would advice for backpropagating all neurons separately in the beginning, and then observe the heatmaps together.

3) If you are using tensorflow/keras, you should have a look at https://github.com/albermax/innvestigate, which was designed with that use case in mind. This very repo right here (.../lrp_toolbox) is quite old and has been built in the age before convenient to use deep learning frameworks. There also is a neat pytorch-solution on the way. In all cases, please read our latest papers [[1](https://link.springer.com/chapter/10.1007/978-3-030-28954-6_10)][[2](https://ieeexplore.ieee.org/abstract/document/9206975)] on how to properly configure and use LRP for deeper models. Configuration and alignment to the model matters for optimal results.

**Comment by xushanthu-2014** [comment]
Hi Sebastian, thanks for your long and patient reply! I will try from the regression one. And by 'out of the box' what do you mean? And based on my understanding, I can just directly apply the regression network with LRP without further masking or categorising? (if one regression value).
In that case I will try it first and then get back to you if I got any issue. Thanks a lot!
